crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -387,7 +391,7 @@ impl<'db> TypeInferenceBuilder<'db> {
 
     /// Are we currently inferring deferred types?
     fn is_deferred(&self) -> bool {
-        matches!(self.region, InferenceRegion::Deferred(_))
+        matches!(self.region, InferenceRegion::Deferred(_)) || self.deferred_state.is_deferred()
```

Comment:
Not sure if I'm missing an issue here, but maybe inferring an `InferenceRegion::Deferred` should set `self.deferred_state` and then this check can be simplified?

You have a note above to add docs for `deferred_state`; I think how these two things relate (and why `InferenceRegion::Deferred` overrides `self.deferred_state`) is a good candidate for some clear docs.

Reply:
Ok after reading the whole diff more carefully, I think I understand this now, but yeah it would be great to document it.